clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -570,6 +604,21 @@ public void requestUpdate() {
             this.metadata.requestUpdate();
         }
 
+        /*
+         * Keep track of any nodes we've ever seen. Add current
+         * alive nodes to this tracking list. 
+         * @param nodes Current alive nodes
+         */
+        private void updateNodesEverSeen(List<Node> nodes) {
+            for (Node n : nodes) {
+                if (!nodesEverSeenById.containsKey(n.id())) {
```

Comment:
I wonder if the id check is sufficient. If a broker was reprovisioned with a new IP address, for example, we'd probably want to update this collection with the new Node.